File: installer/ttpmenu/ttpmenu.pl

#
# Export Tera Term Menu registry to ini file.
#   with ActivePerl
#
# Usage:
# c:\>perl ttpmenu.pl > ttpmenu.ini
#

use Win32::Registry;
use strict;

my $TTMREG = "Software\\ShinpeiTools\\TTermMenu";
my $tips;
my $key;
my @subkeys;

PrintSectionName("TTermMenu");
ExportIniFile($TTMREG);

$HKEY_CURRENT_USER->Open($TTMREG, $tips) or die "Can not open registry";
$tips->GetKeys(\@subkeys);
$tips->Close();
foreach $key (@subkeys) {
#	print "$key\n";
	PrintSectionName("$key");
	ExportIniFile($TTMREG . "\\" . $key);
}


exit(0);


sub PrintSectionName {
	my($name) = @_;
	
	print "[$name]\n";
}


sub ExportIniFile {
	my($path) = @_;
	my($tips);
	my(%vals);
	my($key, $RegType, $RegValue, $RegKey, @bytes);

	$HKEY_CURRENT_USER->Open($path, $tips) or die "Can not open registry";

	$tips->GetValues(\%vals);
	foreach $key (keys(%vals)) {
		$RegType     = $vals{$key}->[1];
		$RegValue     = $vals{$key}->[2];
		$RegKey     = $vals{$key}->[0];
		print "$RegKey=";
		if ($RegType == REG_DWORD) {
			printf "%08x\n", $RegValue;
			
		} elsif ($RegType == REG_BINARY) {
			@bytes = unpack('C*', $RegValue);
			printf "%02x ", $_ foreach @bytes;
			print "\n";
		
		} else {
			print "$RegValue\n";
		}
	}
	print "\n";

	$tips->Close();
}

File: installer/ttpmenu/ttpmenu.py

#
# Export Tera Term Menu registry to ini file.
#   with Python 2.7.6
#
# Usage:
# c:\>python ttpmenu.py > ttpmenu.ini
#

import _winreg
import sys
import struct

TTMREG = "Software\\ShinpeiTools\\TTermMenu";

write = sys.stdout.write

def enum_keys(key):
    try:
        i = 0
        while True:
            yield _winreg.EnumKey(key, i)
            i += 1
    except EnvironmentError:
        pass

def read_keys(key):
    values = []
    for name in enum_keys(key):
    	values.append(name)
    return values


def enum_values(key):
    try:
        i = 0
        while True:
            yield _winreg.EnumValue(key, i)
            i += 1
    except EnvironmentError:
        pass

def read_values(key):
    values = {}
    for name, value, type_ in enum_values(key):
    	write("%s=" % name)
    	if (type_ == _winreg.REG_DWORD):
    		write("%08x\n" % value)
    	
    	elif (type_ == _winreg.REG_BINARY):
    		for s in value:
    			bit = struct.unpack('B', s)[0]
    			write("%02x " % bit)
    		write("\n")
    	
    	else:
    		write("%s\n" % value)
    
    write("\n")
    return


def PrintSectionName(s):
	write("[%s]\n" % s)
	return

def ExportIniFile(regkey):
	key = _winreg.OpenKey(_winreg.HKEY_CURRENT_USER, regkey)
	read_values(key)
	_winreg.CloseKey(key)
	return

def main(s):
	PrintSectionName("TTermMenu")
	ExportIniFile(TTMREG)
	
	key = _winreg.OpenKey(_winreg.HKEY_CURRENT_USER, TTMREG)
	keylist = read_keys(key)
	_winreg.CloseKey(key)
#	print keylist

	for keystr in keylist:
		PrintSectionName(keystr)
		ExportIniFile(TTMREG + "\\" + keystr)
	
	return 0
	
if __name__ == '__main__':
	args = sys.argv
	if len(args) > 1:
		s = args[1:]
	else:
		s = ""
	main(s)


